module/zfs/dmu_objset.c

```diff
@@ -1290,12 +1341,24 @@ dmu_objset_sync(objset_t *os, zio_t *pio, dmu_tx_t *tx)
 		 * We must create the list here because it uses the
 		 * dn_dirty_link[] of this txg.
 		 */
-		list_create(newlist, sizeof (dnode_t),
-		    offsetof(dnode_t, dn_dirty_link[txgoff]));
+		multilist_create(newlist, sizeof (dnode_t),
+		    offsetof(dnode_t, dn_dirty_link[txgoff]),
+		    zfs_arc_num_sublists_per_state,
+		    dnode_multilist_index_func);
 	}
 
-	dmu_objset_sync_dnodes(&os->os_free_dnodes[txgoff], newlist, tx);
-	dmu_objset_sync_dnodes(&os->os_dirty_dnodes[txgoff], newlist, tx);
+	for (int i = 0;
+	    i < multilist_get_num_sublists(&os->os_dirty_dnodes[txgoff]); i++) {
+		sync_dnodes_arg_t *sda = kmem_alloc(sizeof (*sda), KM_SLEEP);
```

Comment:
It would be nice to use a kmem cache for these small short lived allocations.  It may not make a significant performance impact but as a fringe benefit it provides us additional visibility in to this machinery.